miner: timelock the coinbase to the mined block's height

darosior opened the pull request:
The Consensus Cleanup soft fork proposal includes enforcing that coinbase transactions set their
nLockTime field to the block height minus 1, as well as their nSequence such as to not disable the
timelock. If such a fork were to be activated by Bitcoin users, miners need to be ready to produce
compliant blocks at the risk of losing substantial amounts mining would-be invalid blocks. As miners
are unfamously slow to upgrade, it's good to make this change as early as possible.

Although Bitcoin Core's GBT implementation does not provide the `coinbasetxn` field, and mining
pool software crafts the coinbase on its own, updating the Bitcoin Core mining code is a first step
toward convincing pools to update their (often closed source) code. A possible followup is also to
introduce new fields to GBT. In addition, this first step also makes it possible to test future
Consensus Cleanup changes.

The commit making the change also updates a bunch of seemingly-unrelated tests. This is because those tests were asserting error messages based on the txid of transactions involved, and changing the coinbase transaction structure necessarily changes the txid of all tests' transactions.

darosior committed:
miner: timelock coinbase transactions

The Consensus Cleanup soft fork proposal includes enforcing that coinbase transactions set their
locktime field to the block height, minus 1 (as well as their nSequence such as to not disable the
timelock). If such a fork were to be activated by Bitcoin users, miners need to be ready to produce
compliant blocks at the risk of losing substantial amounts mining would-be invalid blocks. As miners
are unfamously slow to upgrade, it's good to make this change as early as possible.

Although Bitcoin Core's GBT implementation does not provide the "coinbasetxn" field, and mining
pool software crafts the coinbase on its own, updating the Bitcoin Core mining code is a first step
toward convincing pools to update their (often closed source) code. A possible followup is also to
introduce new fields to GBT. In addition, this first step also makes it possible to test future
Consensus Cleanup changes.

The changes to the seemingly-unrelated RBF tests is because these tests assert an error message
which may vary depending on the txid of the transactions used in the test. This commit changes the
coinbase transaction structure and therefore impact the txid of transactions in all tests.

The change to the "Bad snapshot" error message in the assumeutxo functional test is because this
specific test case reads into the txid of the next transaction in the snapshot and asserts the error
message based it gets on deserializing this txid as a coin for the previous transaction. As this
commit changes this txid it impacts the deserialization error raised.

darosior committed:
qa: sanity check mined block have their coinbase timelocked to height

Sjors commented:
Tested with Stratum v2 (https://github.com/Sjors/bitcoin/pull/68 + first commit here, SRI pool role connected to their custom signet, sv2 -> sv1 translator and a BitAxe). Indeed I'm seeing blocks at height N with `locktime` values of N-1 and `sequence` of 4294967294 (`0xFFFFFFFE`). I explained below why this just works(tm), unlike with Stratum v1.

There's a question about sequencing here, since it's a preparatory step towards the Great Consensus Cleanup proposal https://github.com/bitcoin/bips/pull/1800.

In a way it's similiar to #30681 and other PRs that prepared our mining code for the timewarp fix from that same BIP. But it's also different. Taking account of the timewarp fix only potentially impacts the timestamp of the first block in a retarget period, i.e. push it a bit in the future. This only happens under circumstances that are common on testnet4 but rare on mainnet. Whereas this PR impacts every mined block.

More specifically, for Stratum v1 this PR changes nothing, because `getblocktemplate` does not echo coinbase `nLockTime`. It will require expanding BIP22 / BIP23 and further changes downstream. But the Stratum v2 Template Provider (see #31098) does pass on this information, so any Stratum v2 pool out there will start doing this once it's in a release.

The fact that Stratum v1 support requires downstream changes is a good reason to get started early. Moreover, because deploying this soft fork is more involved than simply updating Bitcoin Core, it's potentially less safe. By changing the default mining behavior long before the fork activates (or even signals) the actual deployment would be a lot safer.

However the BIP is brand new, so we shouldn't make these (specific) mining changes a fait accompli. Although nothing would break if we later abandon this approach in favor of another BIP30 fix, it still seems better to wait before doing this on mainnet until the BIP has been out there for a while.

Arguably, waiting for the v30 release is long enough and we can simply revert this PR if needed.

Another approach could be to make this testnet only, and then decide if we want to apply it to mainnet a bit closer to the release.

I do think we should have this in v30 to at least enable downstream testing, along with any changes to `getblocktemplate`. And some people may prefer to test on mainnet. So yet another approach could be to put it behind a flag (default true for test networks).

polespinasa reviewed (commented):
Concept ack
Did a simple test with regtest and the blocks are correctly mined with a coinbase tx setting `nLockTime=heigh-1` and `nSequence=4294967294`.

I agree adding this before the soft-fork is good so we give time to miners.
I have one question although, if I'm not wrong the node miner code is no longer used in mainnet so this PR only affects to test networks?

Sjors commented:
> Did a simple test with regtest

> so this PR only affects to test networks

Unfortunately regtest is not representative here. It's the only network for which Bitcoin Core will mine by itself. For all other networks, including testnet (3 and 4), you need to use external software.

polespinasa commented:
> Unfortunately regtest is not representative here. It's the only network for which Bitcoin Core will mine by itself. For all other networks, including testnet (3 and 4), you need to use external software.

This is what I though. So how does this change "help" or "put pressure" to add this new rule to miners software? Wouldn't help more implement the code that will check if a block is following it? Obviously code that would never be ran without another change to add an activation date.

Sjors commented:
> Wouldn't help more implement the code that will check if a block is following it?

That's what activating the soft-fork does. But there's a chicken-egg problem, because one reason to delay activating a soft fork is because part of the ecosystem isn't ready.

> So how does this change "help" or "put pressure" to add this new rule to miners software? 

As I explained in my first comment, for Stratum v2 this just works(tm). For Stratum v1 additional changes are needed in other software. Not sure what the best approach is there. For them the change here is a bit like the proverbial tree in the forest.

Sjors commented on src/node/miner.cpp:
```diff
@@ -158,10 +158,13 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     CMutableTransaction coinbaseTx;
     coinbaseTx.vin.resize(1);
     coinbaseTx.vin[0].prevout.SetNull();
+    coinbaseTx.vin[0].nSequence = CTxIn::MAX_SEQUENCE_NONFINAL;
```
Note to reviewers: `IsFinal()` has been present since the first commit as a way to ignore the sequence number, hence the need to avoid `SEQUENCE_FINAL`.

The proposed consensus rule doesn't require `MAX_SEQUENCE_NONFINAL`, anything other than `SEQUENCE_FINAL` is allowed. Maybe that's good to say in a comment.

darosior replied:
Added a comment.

Sjors commented on src/node/miner.cpp:
```diff
@@ -158,10 +158,13 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     coinbaseTx.vout[0].nValue = nFees + GetBlockSubsidy(nHeight, chainparams.GetConsensus());
     coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
+    Assert(nHeight > 0);
+    coinbaseTx.nLockTime = static_cast<uint32_t>(nHeight - 1);
```
As a sanity check, I changed this to just `nHeight`, which immediately triggers:

```
TestBlockValidity: Consensus::ContextualCheckBlock: bad-txns-nonfinal, non-final transaction
```

(either when calling `getblocktemplate` or in sv2 as soon as the Template Provider tries to build a template)

Dropping the `nSequence` line makes that error go away.

In other words, this guarantees that the coinbase transaction can't have been present in an older block. But as the BIP points out, nLockTime wasn't enforced until block 31,000 or so. 
https://bitcoin.stackexchange.com/questions/90229/nlocktime-in-bitcoin-core

We can then manually verify that all these old blocks have `nLockTime` set to 0:

```sh
bitcoin % for i in {1..31001}
do
bitcoin-cli getblock `bitcoin-cli getblockhash $i` 2 | jq '.tx[0].locktime'
done | grep -v 0
```

Note that any alien reorg shenanigans (now that checkpoints are gone) can't cheat with nLockTime because the rule is now enforced from genesis.

arejula27 replied:
I was discussing with @hodlinator  and @Prabhat1308 that a Coinbase UTXO cannot be spent until 100 blocks have been added after it. Wouldn't it be better to set this to static_cast<uint32_t>(nHeight +99 );? This could simplify the logic for the Coinbase spend condition and potentially make mempool and transaction validation more efficient

polespinasa replied:
> Wouldn't it be better to set this to `static_cast<uint32_t>(nHeight +99 );`?
That would make the transaction invalid and unable to be included in a block until the `nHeight + 99` block is not a condition on when it can be spent, but when it is valid (therefore it can be mined).

arejula27 replied:
True :(

arejula27 replied:
That's true, I confused it with CLTV.

polespinasa commented:
> That's what activating the soft-fork does. But there's a chicken-egg problem, because one reason to delay activating a soft fork is because part of the ecosystem isn't ready

Sorry maybe I didn't explain myself. I meant to introduce the code, but without activating it and without activation date. So if this consensus change doesn't go forward, we make sure that we will not have hard fork because of nodes running a version with the check code implemented. Just put it there to say: "Here it is this new condition, at some point this is going to have an activation date, start implementing it". Maybe that's not the way to do it, just dropping ideas. 

As @darosior said:
> updating the Bitcoin Core mining code is a first step toward convincing pools to update their (often closed source) code

So I think that this is a more "aggressive" (or convincing 😉) way to convince them.

darosior commented:
Rebased.

TheCharlatan commented on test/functional/mining_mainnet.py:
```diff
@@ -63,6 +64,9 @@ def mine(self, height, prev_hash, blocks, node, fees=0):
         block.vtx = [create_coinbase(height=height, script_pubkey=bytes.fromhex(COINBASE_SCRIPT_PUBKEY), retarget_period=2016)]
+        block.vtx[0].nLockTime = 0
+        block.vtx[0].vin[0].nSequence = SEQUENCE_FINAL
+        block.vtx[0].rehash()
```
Just a comment, but I also don't think it would be worth it to re-generate this alternative mainnet chain.

Sjors replied:
Would be good to add a comment here to explain.

We can always regenerate it later.

darosior replied:
Added a comment.

TheCharlatan commented on test/functional/mempool_package_rbf.py:
```diff
@@ -384,7 +384,7 @@ def test_wrong_conflict_cluster_size_parents_child(self):
         package_hex1, _package_txns1 = self.create_simple_package(coin1, DEFAULT_FEE, DEFAULT_CHILD_FEE)
         package_result = node.submitpackage(package_hex1)
-        assert_equal(f"package RBF failed: {child_result['tx'].rehash()} has 2 ancestors, max 1 allowed", package_result["package_msg"])
+        assert_equal(f"package RBF failed: {parent1_result['tx'].rehash()} is not the only parent of child {child_result['tx'].rehash()}", package_result["package_msg"])
```
> The changes to the seemingly-unrelated RBF tests is because these tests assert an error message
which may vary depending on the txid of the transactions used in the test.

This is still confusing to me. Shouldn't this just continue working with the decrementing counter for the sequence number? Or is that entirely unrelated?

darosior replied:
It's only related to the coinbase nSequence/nLockTime change insofar as those change the txid of the coinbase transactions and therefore of all transactions in the test (and all other tests, but this test is one of those tests that implicitly relied on specific txid values).

This test is exercising the error returned by `CheckConflictTopology`, which iterates through the set of conflicts: https://github.com/bitcoin/bitcoin/blob/d91a746815e4428c738f1a096a950292cbdb5469/src/txmempool.cpp#L1277

And returns the first error encountered: https://github.com/bitcoin/bitcoin/blob/d91a746815e4428c738f1a096a950292cbdb5469/src/txmempool.cpp#L1288-L1309

The (ordered) set of conflicts is lexicographically sorted by txid: https://github.com/bitcoin/bitcoin/blob/d91a746815e4428c738f1a096a950292cbdb5469/src/txmempool.h#L396

Therefore, changing nothing but the txid of the transactions involved in the test may cause the order of conflicts checked by `CheckConflictTopology` to change, and another error to be encountered first and therefore the test to fail since it asserts the specific error message.

For example, the check you are commenting on sets up the following topology: two distinct transaction `A` and `A'` are spent by a transaction `B`. `C`, which conflicts with `A`, is broadcast. Here `CheckTopology` could either bail on `C` having two parents ("`C` has 2 ancestors, max 1 allowed") or on `A` having a descendant which has other parents ("`A` is not the only parent of child `C`"). Which error message is returned depends on which is encountered first, which depends on the value of the txid of the transactions involved. :man_shrugging:

darosior replied:
Leaving unresolved as it may be useful to other reviewers.

Sjors commented on src/node/miner.cpp:
```diff
@@ -158,10 +158,13 @@ std::unique_ptr<CBlockTemplate> BlockAssembler::CreateNewBlock()
     CMutableTransaction coinbaseTx;
     coinbaseTx.vin.resize(1);
     coinbaseTx.vin[0].prevout.SetNull();
+    coinbaseTx.vin[0].nSequence = CTxIn::MAX_SEQUENCE_NONFINAL;
```
2c10f93d0bd920f20e91f1be7c82e66af2a52919 nit: can you move this down next to `nLockTime`? This line is what makes the `nLockTime` bit work, so makes sense to have them together. Also, in https://github.com/Sjors/bitcoin/pull/49 I put this behavior behind a startup flag.

darosior replied:
I added the comment as you requested above (https://github.com/bitcoin/bitcoin/pull/32155#discussion_r2018964740), hopefully this makes it clearer.

I don't think there should be a startup flag since i don't see a reason for disabling it: this should not affect miners in any way.

Sjors replied:
It affects any miner that uses the Mining interface, because `createNewBlock` returns a `BlockTemplate` which includes a `getBlock()` method which returns a `CBlock` that contains the coinbase transaction.

Sjors commented on src/test/blockfilter_index_tests.cpp:
```diff
@@ -81,7 +81,9 @@ CBlock BuildChainTestingSetup::CreateBlock(const CBlockIndex* prev,
         CMutableTransaction tx_coinbase{*block.vtx.at(0)};
+        tx_coinbase.nLockTime = static_cast<uint32_t>(prev->nHeight);
         tx_coinbase.vin.at(0).scriptSig = CScript{} << prev->nHeight + 1;
+        tx_coinbase.vin.at(0).nSequence = CTxIn::MAX_SEQUENCE_NONFINAL;
```
Maybe `MakeCoinbase(bool with_lock_time = true)`.

darosior replied:
So i looked into this and i don't think there is a clean way of doing it that would help here. What this function is doing is really a hack, creating a block assembler for a specific tip and mining a block with `prev` that does not correspond to it. However one thing is the `nSequence` doesn't change from one block to another. I added here for clarity but i now think it's unnecessarily confusing, so i removed it.

TheCharlatan replied:
Ah, I see. I also tried to come up with something, but did not land on anything satisfying either. Was briefly thinking of adding an optional field in the `BlockCreateOptions`, but that would defeat the point.

Sjors reviewed (commented):
The churn in the first commit is pretty annoying.

Maybe with a `MakeCoinbase` helper the change can split across more than one commit? E.g. the first commit could set `with_lock_time` to `false` for `ChainType::REGTEST` and `height < 300` so the assume valid blocks don't change.

darosior commented on test/functional/mining_basic.py:
```diff
@@ -362,6 +363,12 @@ def test_block_max_weight(self):
             expected_msg=f"Error: Specified -blockmaxweight ({MAX_BLOCK_WEIGHT + 1}) exceeds consensus maximum block weight ({MAX_BLOCK_WEIGHT})",
         )
 
+    def test_height_in_locktime(self):
```
Looks like this slipped through at some point. Fixed now, thanks.

sedited reviewed (commented):
Concept ACK

darosior commented:
Do people think i should change the value of the coinbase transaction used in the fuzz test here as well? I previously kept it along with the PR introducing validation for those fields, but since i change this in unit tests already it might make sense to do it here.

darosior committed:
qa: timelock coinbase transactions created in fuzz targets

darosior committed:
qa: timelock coinbase transactions created in functional tests

darosior committed:
qa: use prev height as nLockTime for coinbase txs created in unit tests

We don't set the nSequence as it will be set directly in the block template generator in a following
commit.

darosior committed:
contrib: timelock coinbase transactions in signet miner

darosior commented:
I have added a commit to also set nLocktime / nSequence for coinbase transactions created in fuzz targets.

> The churn in the first commit is pretty annoying.

@Sjors (from this and in-person discussions) i have split as much as possible of the changes from this commit into preparatory commits. Now there is in the main commit only the necessary churn to keep commits hygienic. Hopefully it helps.

sedited reviewed (approved):
ACK 895beb30b375c04804e9596ed798ab56730dcae1

ajtowns commented on test/functional/mining_basic.py:
```diff
@@ -30,7 +30,8 @@
     ser_uint256,
-    WITNESS_SCALE_FACTOR
+    WITNESS_SCALE_FACTOR,
+    MAX_SEQUENCE_NONFINAL,
```
Should be sorted

darosior replied:
Done.

ajtowns commented on test/functional/mining_basic.py:
```diff
@@ -362,6 +363,12 @@ def test_block_max_weight(self):
+        self.generate(self.nodes[0], 1, sync_fun=self.no_op)
+        block = self.nodes[0].getblock(self.nodes[0].getbestblockhash(), 2)
+        assert_equal(block["tx"][0]["locktime"], block["height"] - 1)
+        assert_equal(block["tx"][0]["vin"][0]["sequence"], MAX_SEQUENCE_NONFINAL)
```
Maybe add something like

```python
        genblock = create_block(hashprev=int(block["hash"], 16), tmpl=dict(height=block["height"]+1), ntime=block["time"]+1)
        genblock.vtx[0].vin[0].nSequence = SEQUENCE_FINAL
        genblock.vtx[0].rehash()
        genblock.hashMerkleRoot = genblock.calc_merkle_root()
        genblock.solve()
        self.nodes[0].submitblock(genblock.serialize().hex())
        assert_equal(self.nodes[0].getbestblockhash(), genblock.hash)
```

to ensure there's explicit test coverage for blocks with a coinbase whose nSequence is final?

I believe with this PR there remains implicit test coverage via feature_taproot (`gen_test_vectors` uses `SEQUENCE_FINAL`) and via rpc_getblockstats (via the test data in data/rpc_getblockstats.json).

darosior replied:
It feels out of place in mining_basic, and the PR that introduces validation checks for nSequence/nLockTime in coinbase does check various combinations in feature_block, including this one. So i think it's better left there?

ajtowns reviewed (commented):
Approach ACK 895beb30b375c04804e9596ed798ab56730dcae1

darosior commented on src/test/util/mining.cpp:
```diff
@@ -36,10 +38,14 @@ std::vector<std::shared_ptr<CBlock>> CreateBlockChain(size_t total_height, const
+        coinbase_tx.nLockTime = prev_height;
         coinbase_tx.vin.resize(1);
         coinbase_tx.vin[0].prevout.SetNull();
+        coinbase_tx.vin[0].nSequence = CTxIn::MAX_SEQUENCE_NONFINAL; // Make sure timelock is enforced.
```
Sure, done.

TheCharlatan commented on test/functional/test_framework/messages.py:
```diff
@@ -43,6 +43,7 @@
 
 MAX_BIP125_RBF_SEQUENCE = 0xfffffffd  # Sequence number that is rbf-opt-in (BIP 125) and csv-opt-out (BIP 68)
 SEQUENCE_FINAL = 0xffffffff  # Sequence number that disables nLockTime if set for every input of a tx
+MAX_SEQUENCE_NONFINAL = 0xfffffffe  # Highest sequence number that does not disable nLockTime
```
The comment may be confusing in the context of just using nlocktime, so maybe it should be amended to describe its non-BIP68 meaning too?

darosior replied:
Rebased on master.

darosior commented on src/test/util/mining.cpp:
```diff
@@ -44,6 +46,10 @@ std::vector<std::shared_ptr<CBlock>> CreateBlockChain(size_t total_height, const
         coinbase_tx.vout[0].nValue = GetBlockSubsidy(height + 1, params.GetConsensus());
         coinbase_tx.vin[0].scriptSig = CScript() << (height + 1) << OP_0;
+        if (height > 0) {
+            coinbase_tx.nLockTime = static_cast<uint32_t>(height - 1);
```
Good catch, thanks. I should have re-checked with my follow-up PR which introduces validation before pushing the change for the fuzz target. Will do that this time.

darosior replied:
Done. Removed the `height > 0` condition, corrected `height - 1` for `height` and added a comment that the `height` variable does not actually correspond to the block height as it mind be confusing to readers. Verified correctness by rebasing my Consensus Cleanup PR on top of this one.

sedited reviewed (approved):
Re-ACK 3904fcd6680dce7a9a0f5f91c5a3069b9fc81782

jonatack commented:
> I'm a bit reluctant to have this merged while BIP54 is in limbo (https://github.com/bitcoin/bips/pull/1800).

BIP54 draft merged.

sedited reviewed (approved):
Re-ACK a58cb3b1c12c8cb75a87375c50f94c4605bb805d

achow101 commented:
In 8f2078af6a55448c003b3f7f3021955fbb351caa "miner: timelock coinbase transactions"

I'm having a hard time understanding why `rbf_tests.cpp` and `mempool_package_rbf.py` are changed? These are not hardcoded txid changes. Rather the specific changes suggest to me that somehow the semantics of rbf have somehow changed? But that doesn't make sense if the only changes are to the coinbase transaction, why would that affect mempool policy?

achow101 commented:
ACK a58cb3b1c12c8cb75a87375c50f94c4605bb805d

TheCharlatan commented:
> Edit: Oh, is it because of set ordering by txid or something like that, and the error message depends on which tx is checked first? If so, it would be nice to have that in the commit message.

Yes, see https://github.com/bitcoin/bitcoin/pull/32155#discussion_r2052783821 above.
